test/integration/consul-container/test/resource/http_api/client/client.go

```diff
@@ -0,0 +1,359 @@
+		config.Transport = defConfig.Transport
+	}
+
+	if config.HttpClient == nil {
+		var err error
+		config.HttpClient, err = NewHttpClient(config.Transport, config.TLSConfig)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if config.Namespace == "" {
+		config.Namespace = defConfig.Namespace
+	}
+
+	if config.Partition == "" {
+		config.Partition = defConfig.Partition
+	}
+
+	parts := strings.SplitN(config.Address, "://", 2)
+	if len(parts) == 2 {
+		switch parts[0] {
+		case "http":
+			// Never revert to http if TLS was explicitly requested.
+		case "https":
+			config.Scheme = "https"
+		case "unix":
+			trans := cleanhttp.DefaultTransport()
+			trans.DialContext = func(_ context.Context, _, _ string) (net.Conn, error) {
+				return net.Dial("unix", parts[1])
+			}
+			httpClient, err := NewHttpClient(trans, config.TLSConfig)
+			if err != nil {
+				return nil, err
+			}
+			config.HttpClient = httpClient
+		default:
+			return nil, fmt.Errorf("unknown protocol scheme: %s", parts[0])
+		}
+		config.Address = parts[1]
+
+		// separate out a reverse proxy prefix, if it is present.
+		// NOTE: Rewriting this code to use url.Parse() instead of
+		// strings.SplitN() breaks existing test cases.
+		switch parts[0] {
+		case "http", "https":
+			parts := strings.SplitN(parts[1], "/", 2)
+			if len(parts) == 2 {
+				config.Address = parts[0]
+				config.PathPrefix = "/" + parts[1]
+			}
+		}
+	}
+
+	config.Token = defConfig.Token
+
+	return &HttpClient{config: *config, headers: make(http.Header)}, nil
+}
+
+// NewHttpClient returns an http client configured with the given Transport and TLS
+// config.
+func NewHttpClient(transport *http.Transport, tlsConf api.TLSConfig) (*http.Client, error) {
+	client := &http.Client{
+		Transport: transport,
+	}
+
+	return client, nil
+}
+
+// request is used to help build up a request
+type request struct {
+	config *api.Config
+	method string
+	url    *url.URL
+	params url.Values
+	body   io.Reader
+	header http.Header
+	Obj    interface{}
+	ctx    context.Context
+}
+
+// setQueryOptions is used to annotate the request with
+// additional query options
+func (r *request) SetQueryOptions(q *QueryOptions) {
+	if q == nil {
+		return
+	}
+	if q.Namespace != "" {
+		// For backwards-compatibility with existing tests,
+		// use the short-hand query param name "ns"
+		// rather than the alternative long-hand "namespace"
+		r.params.Set("ns", q.Namespace)
+	}
+	if q.Partition != "" {
+		// For backwards-compatibility with existing tests,
+		// use the long-hand query param name "partition"
+		// rather than the alternative short-hand "ap"
+		r.params.Set("partition", q.Partition)
+	}
+	if q.Peer != "" {
+		r.params.Set("peer", q.Peer)
+	}
+
+	if q.RequireConsistent {
+		r.params.Set("consistent", "")
+	}
+
+	if q.Token != "" {
+		r.header.Set("X-Consul-Token", q.Token)
+	}
+
+	r.ctx = q.ctx
+}
+
+// toHTTP converts the request to an HTTP request
+func (r *request) toHTTP() (*http.Request, error) {
```

Comment:
What's the reasoning behind having a separate `request` and not just passing around a `http.Request`?

Reply:
This is because the `request` has use case-specific config such as query params like `namespace`, `partition`, etc.

Reply:
Makes sense! Can we add that as a comment.